### ISSUE TITLE
Unifying label and diary views

### ISSUE THREAD

**Issue by shankari** [e-mission/e-mission-docs#740]
- code for label and diary is in `www/js/diary` and `www/templates/diary`
- diary view is `list.js` and `list.html`
- label view is `inf_scroll.js` and `inf_scroll.html`

In angular1, the UI is in `.html` and the javascript controlling the UI is in `.js` in a type of class called a `controller`

The functionality in both of them is almost the same. Our goal is to modularize and unify the two so that we have one codebase and the users have one place to go do label trips.

**Comment by shankari** [comment]
Task 1 is to take the stuff in the `<ion-item` for the diary or inf_scroll view and create a directive for it. you can pass in the trip as a parameter. There should be very few controls/no controls because most of the user interaction is in the labels which I have already created a directive for.

when you see the `<ion-item`, you will see `class="row"` and `class="col"`. these are part of CSS flow elements which ionicframeworkv1 has defined to ensure that elements can resize as needed.
https://ionicframework.com/docs/v1/components/#grid

**Comment by tyleryandt18** [comment]
Here is the directive that I created
![Screenshot (7)](https://user-images.githubusercontent.com/89606659/175136645-5dd67642-8d14-40a5-a7ef-e6d8e5962129.png)

**Comment by shankari** [comment]
so this is a great start, but it puts too much information into the controller. So if we wanted to use this directive in another screen with its own controller, we would need to copy-paste the `tripSummary` function that combines everything into the return statement into that controller as well. So we are not getting the benefit of pulling it out into a directive.

An even simpler example is if we wanted to have a list of 5 trips that we are showing. Naively you would need to copy `tripSummary` 5 times. Try using `ng-repeat` and convince yourself that the naive approach does not work to show 5 different trips and document what you tried and how it failed so I can know you are understanding it. 

Instead, we want to put the code that creates that string into the directive. We want to have the directive take a trip object with sub-fields for startAddress, endAddress.... and then construct the string inside the directive.

In the e-mission-phone codebase, trip objects are already created by reading them from the server. But you cannot do that yet because you are building this standalone. So you should create a trip object with the fields in an e-mission trip, but populated manually for now. 
e-mission trip object structure: https://github.com/e-mission/e-mission-server/blob/master/emission/core/wrapper/cleanedtrip.py

So the next steps are:
- try to use the current directive to populate a list of 5 different trips and post your attempts here
- change the directive to take a trip object with a similar structure to the e-mission trip object
- populate a list of 5 different trips with the new directive
- (stretch goal) change the HTML for the directive to actually have 2-D structure instead of being a string 
   - you can look at the current ion-item for an example https://github.com/e-mission/e-mission-phone/blob/master/www/templates/diary/list.html#L46

**Comment by tyleryandt18** [comment]
![Screenshot (12)](https://user-images.githubusercontent.com/89606659/175342817-92a300db-1592-48f4-9e09-8f0f42d29e68.png)
![Screenshot (13)](https://user-images.githubusercontent.com/89606659/175342831-487acdbf-6425-4f26-8ef2-6ca51575779e.png)
This was my first attempt to use ng-repeat to create a list of 5 trips. It is a little different than the directive that we were working with earlier because I coded the array of trip objects into it, and therefore with ng-repeat I could print out each attribute (startAddress, endTime, etc). I realize that ng-repeat would not work with the tripSummary function because the trip object itself does not have a tripSummary function, so in the below screenshots when I use ng-repeat="trip in trips", the trip.tripSummary() will not produce anything.  Furthermore, in the directive that we worked on earlier, ng-repeat would not function properly to produce a list of 5 different trips because there is only one trip hardcoded in; only one $scope.startAddress etc. If this isn't the realization you wanted me to come to, let's talk about it in our 1:1.
![Screenshot (14)](https://user-images.githubusercontent.com/89606659/175343857-038d0635-86c8-4dfa-a4cb-1bdd0fcef89b.png)
![Screenshot (15)](https://user-images.githubusercontent.com/89606659/175343879-d8df7dfa-19b4-4ca0-ac0e-e4996c9de61e.png)

**Comment by shankari** [comment]
> If this isn't the realization you wanted me to come to, let's talk about it in our 1:1.

Yup, that is the realization. Note also that with the first approach, you are not even using a directive. You are putting in the `h3` directly into the `ng-repeat`. This is effectively the approach that we use right now in `list.html` only using `collection-repeat` instead of `ng-repeat`.

See if you can read the `list.html` code and map your first solution to it.

Why is this not good enough? Why do we even need a directive?

**Comment by tyleryandt18** [comment]
I might need some guidance going over list.html in our meeting. I do notice the collections-repeat; that part makes sense to me wherein it's taking each trip in data.currDayTripWrappers and displaying it's start time with {{tripgj.display_start_time}}.

I understand that directives allow us to create custom html tags, but what is the benefit of this? I'm not sure why we need a directive other than to consolidate the code from the label and diary views.

**Comment by shankari** [comment]
one of the benefits is the one that you outline right here:

> I might need some guidance going over list.html in our meeting.
> wherein it's taking each trip in data.currDayTripWrappers and displaying it's start time with {{tripgj.display_start_time}}.

As you can see from the app UI, it displays a lot more than the start time, but it is hard to read it in the code because it quickly gets more complex. Pulling out the code into separate modules and sub-modules makes it easier to read and modify one small part of a complex system.

**Comment by tyleryandt18** [comment]
> As you can see from the app UI, it displays a lot more than the start time, but it is hard to read it in the code because it quickly gets very complex. Pulling out the code into separate modules and sub-modules makes it easier to read and modify one small part of a complex system.

Ok, yeah that makes sense. Using directives to make the code easier to read / modifying smaller parts of a complex system

**Comment by tyleryandt18** [comment]
Here's my attempt to populate a list of trips using directives. Again, I know it is very simple, but I've never worked with directives before and I feel like it's a good baby-step start
![Screenshot (16)](https://user-images.githubusercontent.com/89606659/175355001-5b8d415c-ca98-4da8-b459-02c78a16b73e.png)
![Screenshot (17)](https://user-images.githubusercontent.com/89606659/175355021-33b7f29a-97d8-497d-9b64-672932826
![Screenshot (18)](https://user-images.githubusercontent.com/89606659/175355028-63d722aa-e3c7-4e3d-aff2-c5ecdcead7b9.png)
115.png)
.

**Comment by shankari** [comment]
concretely, for step 1, the goal is for your HTML to look like (psuedo-code, DO NOT COPY AND PASTE).

```
<div ng-repeat="trip in tripList">
<trip-summary trip="trip"></trip-summary>
</div>
```

and have the output be the same as your prior output.

You literally have to create a directive that encompasses everything within the `h3`

**Comment by tyleryandt18** [comment]
- look up how to pass an object into a directive
- create 5 trips with that directive without repeating the functions like in the code that I have here already
- 
- look up creating two-dimensional structures for directives
- install and log trips into CanBikeCO to see what is going on in the list.html

**Comment by tyleryandt18** [comment]
![Screenshot (22)](https://user-images.githubusercontent.com/89606659/175761165-f0f68e8f-3e8a-4a0d-bf13-ff65521de538.png)
![Screenshot (21)](https://user-images.githubusercontent.com/89606659/175761167-10ce2af2-3dfa-467b-8a27-f6dc76fb02a0.png)
![Screenshot (20)](https://user-images.githubusercontent.com/89606659/175761169-f37d1541-595a-4c3d-ac38-a4dc8ca6f99f.png)
![Screenshot (19)](https://user-images.githubusercontent.com/89606659/175761170-094ebc40-6f39-427e-9b24-62907955ee8f.png)
Here is what I was able to get through on Friday. I understand now how to pass an object into a directive, but I did have struggles with populating the list within the directive, and not just simply having the list of 5 trips already established in the `ng-init`. I tried to create a function within the directive that would push a trip into a `tripList ` array. This `tripList` array would initially be an empty array, which could be accessed the same way that we are accessing each trip `tripList: '=tripList'`, and then the function would be `function addTrip(trip, tripList) {tripList.push(trip)};`, but I couldn't get it to work.

I was unable to format a 2D structure. I did not have enough time over Friday, nor do I fully understand the list.html elements even after logging trips on the CanBikeCo app.

**Comment by shankari** [comment]
so this is actually fine so you are done with everything other than the 2-D structure.
My high level comments are:
- naming matters: so don't call this `my-directive`
- the template is typically not a full HTML document, but the part that is within the body. So the top level of the directive would be something like `<ion-content` or `<div` or something like that.
Concrete examples:
- https://github.com/e-mission/e-mission-phone/blob/master/www/templates/survey/multilabel/multi-label-ui.html (top level is a `div`)
- https://github.com/e-mission/e-mission-phone/blob/master/www/templates/metrics/range-display.html which just has a span

**Comment by shankari** [comment]
- Use only iOS for now
- `ios-sim` or `xcrun simctl` to start a simulator
- unzip the app.zip and drag/drop to the simulator to get the devapp working (check with @sebastianbarry if you have issues)


Next task:
- **Get the app working on your laptop and be able to load data**
   - If you don't want to set up a server, you can use the staging server with your token but make sure not to mess up the data too badly because we do use it for analysis down the road.
  - To connect to the staging server, create a file in the phone repo called `www/json/connectionConfig.json` with the following text
```
{
    "connectUrl": "http://stage.canbikeco.org",
    "aggregate_call_auth": "no_auth",
    "android": {
        "auth": {
            "method": "prompted-auth",
            "clientID": "ignored"
        }
    },
    "ios": {
        "auth": {
            "method": "prompted-auth",
            "clientID": "ignored"
        }
    }
}
```
and login with your token

Next step after that:
- Pull out everything in the `ion-item` in the infinite scroll (https://github.com/e-mission/e-mission-phone/blob/master/www/templates/diary/infinite_scroll_list.html) into a directive that you pass the trip into.
  - The HTML part of this is relatively easy
  - The javascript should also be fairly easy because you don't have to touch the code around how the trips are read, you just need to pass in correctly from the `ng-init`

Both of these done in 2 days (by Wed evening).

**Comment by tyleryandt18** [comment]
Pulled out the ion-item into a directive. Instead of using the scope, simply used the itemHt and passed in a height of 150, which was established in the $scope.

**Comment by shankari** [comment]
the diary and infinite scroll directives are almost the same visually, but they expect the trip structure to be fairly different.

#### diary
in the diary items, each trip is expected to be a geojson object. geojson is a standard format to represent geographic data as json.https://geojson.org/

So it retrieves the full list of trips for a particular day, with each trip as geojson that includes all of the objects that make up the trip. So I think that the trip is a geojson feature collection: start feature, end feature, _n_ section features, each section is a feature collection with a start/stop and a linestring for the trajectory. Each feature has properties for distance/duration/...

Putting a breakpoint at https://github.com/e-mission/e-mission-phone/blob/master/www/js/diary/list.js#L204
will allow you to see the base structure of the geojson that is retrieved from the server.
`DiaryHelper.directiveForTrip);` adds some formatting fields to the retrieved trip geojson. Again putting a breakpoint at
https://github.com/e-mission/e-mission-phone/blob/969d8a1274eda809e261506f43d7cc6540d036df/www/js/diary/list.js#L211
will help you see what new fields are added and map them to what is shown in the directive.

#### infinite scroll

For the infinite scroll, the retrieved trips are in the openpath trip format. 
https://github.com/e-mission/e-mission-server/blob/master/emission/core/wrapper/confirmedtrip.py

Note that this does not have sections or trajectories embedded when we retrieve them. We retrieve a downsampled version of the trajectory when the map popup is opened, which is why we don't show the map by default.

To see the trip format here, put a breakpoint at:
https://github.com/e-mission/e-mission-phone/blob/969d8a1274eda809e261506f43d7cc6540d036df/www/js/diary/infinite_scroll_list.js#L108

to see how we retrieve the trajectory

that returns the trip as geojson with an embedded trajectory, which we can then show in a leaflet map
https://github.com/e-mission/e-mission-phone/blob/969d8a1274eda809e261506f43d7cc6540d036df/www/js/diary/infinite_scroll_detail.js#L70

```
        <leaflet geojson="tripgj" id="infscroll-detail" defaults="mapCtrl.defaults" 
            height="50%" width="100%" data-tap-disabled="true">
```

This is particularly confusing because we sometimes call a variable `tripgj` aka `trip geojson` when it is actually a trip object.

to unify the two, we should really standardize on one method for retrieving and representing trips, otherwise, we will continue to have two implementations and a lot of confusion.

**Comment by tyleryandt18** [comment]
I am going to go ahead and go out on a limb here and say that we should implement the diary trip item, the one that is a geoJSON. As you noted in this second-to-last line, there is a confusion when we read in the trip from inf_scroll as a tripgj. Furthermore, I feel as if having the map data in the trip without having to do that extra step of retrieving the trajectory when we open the map popup.

NOW
If I am wrong, or if I have the same implementation as you do in mind but the wrong reasoning behind it, or if I have let out any key details, please let me know. I would love to understand this fully, and as it sits right now I am still a little shaky on my understanding. I also believe that the structure of the diary tripgj was much simpler
<img width="961" alt="Screen Shot 2022-07-09 at 7 42 13 PM" src="https://user-images.githubusercontent.com/8
<img width="961" alt="Screen Shot 2022-07-09 at 7 43 45 PM" src="https://user-images.githubusercontent.com/89606659/178128941-4437a360-d5b9-4562-b593-121d40f85425.png">
9606659/178128940-93d3471e-6862-4439-b306-0245687d46d0.png">
.

**Comment by shankari** [comment]
> Furthermore, I feel as if having the map data in the trip without having to do that extra step of retrieving the trajectory when we open the map popup.

Agreed that having the map inline instead of opening another popup is useful.

> I also believe that the structure of the diary tripgj was much simpler

Not sure why this is; can you clarify?

> I am going to go ahead and go out on a limb here and say that we should implement the diary trip item, the one that is a geoJSON. 

Did you look into which came first, diary or label screen? What were the problems with the older version that prompted the creation of a newer version? Did you search through the existing list of issues?

**Comment by tyleryandt18** [comment]
@shankari 
> Not sure why this is; can you clarify?
<img width="839" alt="Screen Shot 2022-07-09 at 8 11 14 PM" src="https://user
<img width="961" alt="Screen Shot 2022-07-09 at 7 43 45 PM" src="https://user-images.githubusercontent.com/89606659/178159542-cd4025d9-1dd8-4a32-af0d-79a41840040f.png">

<img width="839" alt="Screen Shot 2022-07-09 at 8 11 14 PM" src="https://user-images.githubusercontent.com/89606659/178159609-bc58edb2-1cc1-4cb6-89d3-d71002a79cba.png">


When looking at the trip objects at those breakpoints, the trip in the diary has less features. I believe this is due to the fact that in the diary item (top screenshot), the start time and end time and trip information is saved in the geoJSON, and therefore we do not need to create the separate fields such as in the trip object in the inf_scroll (bottom screenshot).

> Did you look into which came first, diary or label screen? What were the problems with the older version that prompted the creation of a newer version? Did you search through the existing list of issues?
It appears that the list.js appeared first, as the oldest commit change that I noticed was 6 years ago as opposed to the inf_scroll which was 2 years ago.

I found this issue particularly interesting, #623. Idk if that links to the issue, but the issue stated that the start and end times could not be read into the leaflet map from the tripgj, so a start and end time attribute were added, possibly causing the implementation of the trip object that we know in the inf_scroll.

**Comment by shankari** [comment]
Do you know how to find the original checkin for various files? Look at Github blame - e.g. 
https://github.com/e-mission/e-mission-phone/blame/master/www/js/diary/infinite_scroll_list.js
and
https://github.com/e-mission/e-mission-phone/blame/master/www/js/diary/list.js

A search for prior diary issues could also be instructive
https://github.com/e-mission/e-mission-docs/search?p=3&q=diary+screen+load&type=issues

Notably: https://github.com/e-mission/e-mission-docs/issues/226 and 
https://github.com/e-mission/e-mission-docs/issues/488 and
https://github.com/e-mission/e-mission-docs/issues/707

**Comment by tyleryandt18** [comment]
Ok so looking over those issues, loading in the data from the geoJSON trip is too slow. This is due to the fact that loading in the trip onto the map in the geoJSON takes too long? That checks out in my mind, mapping each trip onto the map in the diary screen would take some time. And yes, I was using the blame feature, but perhaps in the wrong way.

So, with this delay in loading in the data, we should use the trip object found in the inf_scroll_list. This would provide a solution to the slow loading of the geoJSON trip since the map leaflet would only appear when we go to the popup, ie click on the trip on the label screen.

**Comment by shankari** [comment]
@tyleryandt18 yes, I think we should use the trip object found in the inf_scroll_list. However, that provides much less information than the diary screen. Note that the bug wants to focus on _unifying_ the list and diary screens, not just picking one.

- We should think through which aspects for the diary screen we want to port over to the label screen
  - we probably do want the inline map, BUT
  - we may not want to port everything over - maybe the diary screen is too complex the way it is
  - you may want to brainstorm with @sebastianbarry/other interns/friends and family on what the new unified design should have and draw it out on figma et al if needed.
- Then, we can figure out how to populate those aspects by using the base e-mission objects

**Comment by tyleryandt18** [comment]
@shankari 
Ok great! Noted on the unification, not just picking one.

I'm actually looking over right now what features I would want to pull over from the diary screen as you sent this! I think the in-line map is a must. I also agree with @sebastianbarry regarding the color scheme of the UI, that NREL blue looked really good.

**Comment by tyleryandt18** [comment]
@shankari
High-Level Implementation Functionality for the Unified Screen
- in-line map leaflet
- left-to-right scroll instead of infinite up-down. I feel like this will provide a simpler look because you are only seeing one trip at a time, therefore seeing less information at once.
- NREL blue replacing the mint green
- include the percentages of trip mode on the trip
<img width="384" alt="Screen Shot 2022-07-13 at 10 39 10 AM" src="https://user-images.githubusercontent.com/89606659/178786301-0821e0be-9e54-411b-8ed8-d5816c7df4ac.png">

Other than that, I think that our infinite_scroll screen has a good baseline. It's simple and concise, I think that maybe we just switch to a left to right scroll, sort of like the "notebook feel" that @sebastianbarry had mentioned. Or, maybe we implement the diary screen's method of having the date drop-down, and then you can scroll through your trips up-down for that given day.

**Comment by tyleryandt18** [comment]
Instead of the left-right scroll, let's go with the diary screen's date chooser, where you can go to a certain day and then see all of the trips from that day.

**Comment by tyleryandt18** [comment]
@shankari 
I was thinking that it would look the same as the current diary view, but I think that the current issue with the slow retrieval is in relation to this date selector; once the user chooses a date, the app then must retrieve the data from the server. Could we fix this issue by storing the data locally on the user's phone?

**Comment by shankari** [comment]
@tyleryandt18 the slow retrieval is not the only issue; please read through all the related issues that I have tagged.

A challenge with the date selector approach is that you have to know which is the previous date on which:
- you had a trip
- you had an unlabeled trip

How would you propose to overcome that challenge?

**Comment by tyleryandt18** [comment]
@shankari
I see what you mean by this issue. That is a nice feature of the infinite scroll Label screen, that it can show you the trips that you need to label.

A solution that I have for this is that we could implement a notification, perhaps a red flag, on the dates that you have not yet labeled a trip for. And when you are on a date that you have all your trips labeled, there could just be a little red flag on the date itself so you know that you have trips to label.
<img width="319" alt="Screen Shot 2022-07-14 at 2 50 12 PM" src="https://user-images.githubusercontent.com/89606659/179082924-8d1dd990-bdc6-47f0-8f3c-cb8382c2bb89.png">

For example, the red flag would appear in the corner of the "11 Jul" button on the top if I had a trip to label for the 10th.


Furthermore, in terms of aesthetics, I find their simple color scheme super appealing. I think we should stick with one color, the NREL blue throughout. I think it looks good that their buttons are not red when unlabeled. For us, the purpose and mode button are red if you haven't labeled a trip. Although I like that aesthetic choice, it does provide functionality in the sense that it alerts the users that they haven't labeled a trip there. My thought is to leave the trips white as our friends in Montreal have, but carry over the red flag from earlier and attach it to a label button that has not yet been filled out.

I also like the split down the middle with the map leaflet on the left and trip data on the right, but I wouldn't want to copy their design choices.

**Comment by shankari** [comment]
@tyleryandt18

First, I want you to stop thinking of the final solution as either/or. Our goal is not to pick either infinite scroll or diary but to unify them. What might a unified view look like?

>  I see what you mean by this issue. That is a nice feature of the infinite scroll Label screen, that it can show you the trips that you need to label.
> A solution that I have for this is that we could implement a notification, perhaps a red flag, on the dates that you have not yet labeled a trip for. And when you are on a date that you have all your trips labeled, there could just be a little red flag on the date itself so you know that you have trips to label.

I am going to push back on this for two reasons:
1. The infinite scroll not only shows you trips that you have to label, but also "All trips" as an unbroken timeline that you don't have to keep clicking through for. Let's say I have labeled all my trips but I want to scroll through to what I did last week. With diary, you would need to click though to last Sunday and then click through day by day. And on some days, I wouldn't even have any trips. Why is more clicks better than fewer clicks?
1. You also need to think about ease of implementation. How concretely would you determine (from the phone) which dates have trips to label? How and where would you display the flag so users could easily see it?

> For example, the red flag would appear in the corner of the "11 Jul" button on the top if I had a trip to label for the 10th.

What if you had labeled 9th and 10th, but had an unlabeled trip for the 8th? You launch the app and you see the diary for the 11th. What do you do?

The diary option is essentially an query pattern for trips - pulling out the trips for one particular day. Can you articulate your reasons why you want to pick this, and *only this* access pattern? I haven't yet heard a reason why supporting only day-by-day access is optimal.

> Furthermore, in terms of aesthetics, I find their simple color scheme super appealing. I think we should stick with one color, the NREL blue throughout. I think it looks good that their buttons are not red when unlabeled. 

This is because they haven't implemented the label assist feature yet.

> For us, the purpose and mode button are red if you haven't labeled a trip. Although I like that aesthetic choice, it does provide functionality in the sense that it alerts the users that they haven't labeled a trip there.

Note that although you don't have them, we also have yellow labels to indicate our guesses for the labels. You can see them too if you actually label your trips!

>  My thought is to leave the trips white as our friends in Montreal have, but carry over the red flag from earlier and attach it to a label button that has not yet been filled out.

What would this look like concretely (sketch plz)? And how would you deal with yellow labels?

**Comment by tyleryandt18** [comment]
> I am going to push back on this for two reasons:
> 
> 1. The infinite scroll not only shows you trips that you have to label, but also "All trips" as an unbroken timeline that you don't have to keep clicking through for. Let's say I have labeled all my trips but I want to scroll through to what I did last week. With diary, you would need to click though to last Sunday and then click through day by day. And on some days, I wouldn't even have any trips. Why is more clicks better than fewer clicks?

You're right, I wasn't thinking of the number of clicks, more in the sense of organization by date. I was thinking more along the lines of what if I wanted to view a trip from three months ago? (not sure why I would per say, so maybe that is an important note too) With the infinite scroll screen, I would have to scroll up all the way to find that trip, as opposed to the date selector, where I can go to that certain date and look at it with ease.

> 2. You also need to think about ease of implementation. How concretely would you determine (from the phone) which dates have trips to label? How and where would you display the flag so users could easily see it?
In the multiabel html, we have an ng-if determining if it's labeled, inferred, or neither. Very elementary, but couldn't we include the flag popup if.... ng-if="!trip.userInput[input.name] && !trip.finalInference[input.name]"? And include the flag icon in the button class?

> > For example, the red flag would appear in the corner of the "11 Jul" button on the top if I had a trip to label for the 10th.
> 
> What if you had labeled 9th and 10th, but had an unlabeled trip for the 8th? You launch the app and you see the diary for the 11th. What do you do?

There would be a flag icon in the date selector. I will draw a sketch of what I mean by this. You then would click on the date selector, see the calendar, where there would be a flag indicator on the date with an unlabeled trip, then go to that date and label it. But you're right, after typing that out, the implementation doesn't sound too easy.
> The diary option is essentially an query pattern for trips - pulling out the trips for one particular day. Can you articulate your reasons why you want to pick this, and _only this_ access pattern? I haven't yet heard a reason why supporting only day-by-day access is optimal.


I believe that the main reason why I was in favor of that access pattern is for accessing past trips. With the infinite scroll access pattern, I have to scroll up to find a previous date. My mind fixated on this problem without considering that that issue may not arise as often as I think, because how often do users access previous trips? Probably not often after they are already labeled.

> > For us, the purpose and mode button are red if you haven't labeled a trip. Although I like that aesthetic choice, it does provide functionality in the sense that it alerts the users that they haven't labeled a trip there.
> 
> Note that although you don't have them, we also have yellow labels to indicate our guesses for the labels. You can see them too if you actually label your trips!


I do see what you mean by the yellow labels. I think stylistically we should consider a different color scheme instead of the red green and yellow. Aesthetically I think they clash, but in terms of functionality, they provide good indicators.

I will work on the sketches now.

**Comment by tyleryandt18** [comment]
<img width="322" alt="Screen Shot 2022-07-14 at 4 16 47 PM" src="https://user-images.githubusercontent.com/89606659/179099165-06a4b1a0-5aaa-420d-8e43-4d26ff1d2d47.png">

Red flag notifying that there is a day without a trip labeled




<img width="322" alt="Screen Shot 2022-07-14 at 4 28 06 PM" src="https://user-images.githubusercontent.com/89606659/179099192-5fa1e23b-d437-4a0f-b704-b6b5e895abe1.png">

Red flag indicating that the day to label is on the fifth

**Comment by tyleryandt18** [comment]
But again, I see the benefit of the infinite scroll access as opposed to date selector (I understand that we are creating a unified screen, that is just poor wording), but I wanted to diagram my thoughts for how the red flag idea would work.

**Comment by shankari** [comment]
> You're right, I wasn't thinking of the number of clicks, more in the sense of organization by date. I was thinking more along the lines of what if I wanted to view a trip from three months ago? (not sure why I would per say, so maybe that is an important note too) With the infinite scroll screen, I would have to scroll up all the way to find that trip, as opposed to the date selector, where I can go to that certain date and look at it with ease.

You are absolutely right that the infinite scroll pattern also has its drawbacks, which is why I think we should come up with a way to **unify the two approaches** and support **both kinds of access patterns.**

Think through how the other apps handle the use case that you want to be able to scroll smoothly and see your history/timeline but also jump around in it to specific places?
- Just checked and WhatsApp doesn't do anything specific - you do just have to scroll back to find older messages, even if they are from a year ago (so we have to be better than WhatsApp. Hm....)
   - WhatApp does have a search button, though. Maybe thinking of this as a searching problem can give you some ideas.
   - For example, what if we just added a date selector to the top of the infinite scroll screen? People can then use it to quickly "jump" to a place in the scroll list, but then continue scrolling from there to avoid clicks.
- I don't really have any other social media (I am not on FB/IG/etc) to use as examples, but maybe you or your friends/family do 
- In a related domain, how do calendar apps (which have the same problem) deal with this? at least on mobile, you want to be able to see your meetings for a week "at a glance" but you want to be able to pick a date as well.

**Comment by tyleryandt18** [comment]
@shankari
I'm thinking that on our unified screen, we add a button to the top that has a calendar popup that can take you to a certain day. I liked your idea above of just adding a date selector to the top of the infinite screen. I am thinking that it would look something like this, with the to label/all trips/unlabeled trips buttons shifted over to the right slightly more to create even spacing.
Then the date selector would function the same, with the red flag also appearing to mark days that you haven't labeled yet.

<img width="279" alt="Screen Shot 2022-07-15 at 10 50 05 AM" src="https://user-images.githubusercontent.com/89606659/179272583-b19a052c-cd2e-4550-a51a-aff85f8c6875.png">

**Comment by shankari** [comment]
@tyleryandt18
- we could also consider removing "All Unlabeled" or convert the labeling buttons to a dropdown to make some room. Not sure how many people are using each screen, maybe check with the focus group as well?
- also what color do you plan to use for the red labels? you should have a screenshot showing them as well.

Plan out which 3 options (colors/layout) you are going to mock up and share with the other interns for feedback

**Comment by tyleryandt18** [comment]
wrt the red labeling, I was planning on having a white button with the NREL as an outline for it, and then once the user labels it switches over to the NREL blue full button. I think that having All Trips as the default is a good idea. I'm not sure what the difference is between unlabeled and to label trips is, so maybe we eliminate that and only have the three filters; Calendar, To Label, and All Trips.

I will look into the calendar library.

**Comment by shankari** [comment]
> Plan out which 3 options (colors/layout) you are going to mock up and share with the other interns for feedback

I would first mock up and get feedback

**Comment by shankari** [comment]
Also, think about the labels to make more sense:
- mode -> dominant mode?
- purpose -> destination?
- replaced mode -> if unavailable?

**Comment by shankari** [comment]
> I'm thinking that on our unified screen, we add a button to the top that has a calendar popup that can take you to a certain day. 

Could you also list out what other apps do as a sort of "related work"? At least in a lab, we would like to have more justification for our decisions than what one person likes.

**Comment by tyleryandt18** [comment]
@shankari

The only social media platforms that I have are instagram and a fairly new format called BeReal. Both have implementations of an infinite scroll on the feed. I could provide screenshots if you would like, but I would have to import from my phone. I suggest that we scratch the calendar feature. I can't find any social media platforms to support implementing that feature. Nor would I think that one would want to see their past trips with enough enthusiasm to the point in which it is worth implementing, if that makes sense. If you want to see a past post on instagram, you either have to scroll until you find it, or search for a person's profile to find that user, and then scroll on their specific profile's feed to find that post in the past.

My mockups have only two filters on the screen, All Trips and To Label. This provides the users to have an infinite scroll feature to see all of the trips that they need to label, and then a scroll to see all of their past trips. I received feedback from the team on the color schemes, so the ones that I have implemented here are the most popular. I have not yet shown them these mockups, so I will send out some surveys to them after I finish writing this up.

Mockup 1: I figured I would just include to the side what an inferred label color would look like, because adding in another trip would extend past the visible screen on a phone anyway.
<img width="510" alt="Screen Shot 2022-07-18 at 1 33 36 PM" src="https://user-images.githubusercontent.com/89606659/179605404-0914a061-ff69-4764-8745-670ce6746480.png">

Mockup 2:
<img width="512" alt="Screen Shot 2022-07-18 at 1 39 04 PM" src="https://user-images.githubusercontent.com/89606659/179605513-707c8553-2e50-497e-96cc-7f56f3d4b33f.png">

Mockup 3:
<img width="288" alt="Screen Shot 2022-07-18 at 1 44 20 PM" src="https://user-images.githubusercontent.com/89606659/179605555-a65fdd1b-4207-4eca-95d5-93ddead3e7cb.png">

My Figma skills are still very elementary, so I understand that not everything lines up to an exact pixel, but obviously that will be fixed when I decide upon a design and implement it with code.

Again, popular social media supports the decision of an infinite scroll, as it is found to be very addictive. A recent trendy app is TikTok. I don't use it because it is so addictive. It uses an infinite scroll feature to display its content. Recently, it has even begun to develop updates to stop you from infinite scrolling because it has been so addictive. I'm not saying that our users will become addicted to labeling their trips, but it is a better option as opposed to the calendar / date selector.

**Comment by tyleryandt18** [comment]
wrt to the interns' opinions on red/yellow/green vs new color scheme, most of them stated that they preferred the new color schemes, as it provided a clean look. I feel as if we take their feedback in creating a tutorial/color schematic key, and it will provide an easier usage.

**Comment by shankari** [comment]
> Maybe we should implement a tutorial walkthrough.

There is already one for the diary view. You should adapt it for the unified view.

> Other interns also suggested that there be a color-scheme key whether that be with the blue scheme

Where would we put this?

> A common theme that I have seen in all feedback forms is that there is no motivation to label trips, and it feels like a chore to do so.

Well, if you can figure out how to make that fun, I can pretty much guarantee you a publication.

> AngularJS has a date selector/calendar library that supports notifications, it "can display events on a month, week or day view. It's called angular-calendar.

Can you add a link to it? The only one I could find was https://www.npmjs.com/package/angular-calendar which is for angular 12+. We are on angular1.

> > I also don't see any check on how calendar apps (which do care a lot about looking up dates) work.

Any results of the investigation into how calendar apps work?

You should send out the mockups for feedback from the current set of interns, and in parallel, we should plan out how to proceed with the code changes so we don't get too stuck on the UI.

**Comment by tyleryandt18** [comment]
> 
> There is already one for the diary view. You should adapt it for the unified view.
> 
> > Other interns also suggested that there be a color-scheme key whether that be with the blue scheme
> 
> Where would we put this?
> 

When I port the tutorial from the diary view into the unified view, perhaps I could add a color schematic into the tutorial. Just a thought.

> > A common theme that I have seen in all feedback forms is that there is no motivation to label trips, and it feels like a chore to do so.
> 
> Well, if you can figure out how to make that fun, I can pretty much guarantee you a publication.
> 

The first idea that I had was that you could have "friends" on the app, pulling that idea from virtually any social media platform. But after thought, we probably couldn't share others travel data with others, if that makes sense.
> > AngularJS has a date selector/calendar library that supports notifications, it "can display events on a month, week or day view. It's called angular-calendar.
> 
> Can you add a link to it? The only one I could find was https://www.npmjs.com/package/angular-calendar which is for angular 12+. We are on angular1.
> 

Yes, that is the one I was talking about. I will see if I can find one for angular1. 
> > > I also don't see any check on how calendar apps (which do care a lot about looking up dates) work.

https://github.com/angular-ui/ui-calendar
https://angular-ui.github.io/ui-calendar/
I'm not sure if we could use this, it doesn't explicitly state whether it functions on Angular1.

StackOverflow pointed me in the direction of this library, where a user stated that they used it in their Angular1 project https://github.com/mattlewis92/angular-bootstrap-calendar.
> 
> Any results of the investigation into how calendar apps work?
> 
I looked into the Angular1 library linked above. He has functions to create events for a day, week, month or year. And then in the html template, the functions are called so that you can see any events that you have in that specific period of time(year, day, week, month). I haven't ran the code yet, but I will do that tomorrow to look into the workings of the JS/html. 
> You should send out the mockups for feedback from the current set of interns, and in parallel, we should plan out how to proceed with the code changes so we don't get too stuck on the UI.

I've sent out the mockups to all except I think Sebastian, Kacie, and Zach. But I will do that tomorrow morning as well.

**Comment by shankari** [comment]
@tyleryandt18 let's start making progress on the hard-to-code parts that we know we will need to implement while waiting to hear back from interns, friends and family, focus group...

#### Inline map
We don't want to make the mistakes of the original diary and pull all the trajectory information at the same time and slow down the loading
- we keep the current approach to loading the map (which loads a subsampled version of the trajectory and can maybe even drop down the subsampling since the map is smaller) BUT
  - we do this to "fill in" the map, similar to filling in the address after the initial load of the trips
- another option that we have considered is to have the map created as an image on the server so we don't download the map tiles and the map trajectory separately. instead we create and save the map for a trip as an image on the server and then just pull the image when we need it
   - we can create these images for the last 10 days or something like that and keep rotating them and removing the older entries OR
   - we can just create the image when we process the trip and save it as a "thumbnail" for ever

**Comment by shankari** [comment]
@tyleryandt18 Note that, as part of the unification, we also want to make sure that the label screen displays draft trips. That will require a fair amount of javascript refactoring, so let's focus on getting the maps inline for processed trips first and then start on the draft trip display.

**Comment by shankari** [comment]
@lgharib @PatGendre @asiripanich @jf87 since we are changing the visual representation of anyway while unifying the diary and label screens, we thought we might as well improve it. We were also inspired by the Ma Mobilite redesign which looks a lot cooler than the current UI.

@tyleryandt18 has created these mockups, please vote for which you like best! 

| Mockup 1 | Mockup 2 | Mockup 3 |
|------|------|------|
| <img width="510" alt="Screen Shot 2022-07-18 at 1 33 36 PM" src="https://user-images.githubusercontent.com/89606659/179605404-0914a061-ff69-4764-8745-670ce6746480.png"> | <img width="512" alt="Screen Shot 2022-07-18 at 1 39 04 PM" src="https://user-images.githubusercontent.com/89606659/179605513-707c8553-2e50-497e-96cc-7f56f3d4b33f.png"> | <img width="288" alt="Screen Shot 2022-07-18 at 1 44 20 PM" src="https://user-images.githubusercontent.com/89606659/179605555-a65fdd1b-4207-4eca-95d5-93ddead3e7cb.png">

**Comment by asiripanich** [comment]
> <img alt="Screen Shot 2022-07-14 at 4 16 47 PM" width="322" src="https://user-images.githubusercontent.com/89606659/179099165-06a4b1a0-5aaa-420d-8e43-4d26ff1d2d47.png">

Why is this not an option?

**Comment by shankari** [comment]
> Why is this not an option?

That is pretty close to Mockup #2, right? The main difference is the base color theme which is "NREL blue" 😄  for obvious reasons

**Comment by asiripanich** [comment]
Fair enough. However, I like that one better because of the position of the start and end times.

**Comment by PatGendre** [comment]
@shankari I agree with @asiripanich , the position of the start/times is clearer on the mockup2 variant :-)
Also, it shows a date selector : I think in this issue you want to solve in a generic fashion both the current diary and label (infinite list) views so maybe this is why you don't show explicitly the date in your mockps, however in both cases the date should appear clearly somewhere nearby the trip, in my opinion.

**Comment by tyleryandt18** [comment]
@PatGendre @asiripanich @shankari
Hello all, I'm working on the unification of NREL's diary and label screen. These mockups reflect a short period of time where I was considering removing the date selector feature altogether, but I can assure you that we will include it, after some thought it seems necessary. I was considering have a filter tab in which you could jump to a specific date from a calendar popup. The final design I'm envisioning will look something along the lines of:
<img width="265" alt="Screen Shot 2022-07-20 at 10 53 21 AM" src="https://user-images.githubusercontent.com/89606659/180039302-7c235e62-d4b4-4ed4-93aa-e61cfb91bc87.png">

And yes, as Shankari stated my Figma skills are still pretty elementary haha, that's why you can see some shadows here and there and the buttons aren't lined up perfectly.
<img width="215" alt="Calendar Popup" src="https://user-images.githubusercontent.com/89606659/180039519-99280d40-a8aa-47bd-b2f7-9cb56aac87d7.png">
So my thought process is that when you select the calendar filter, this popup will open and allow you to choose a specific date.

**Comment by shankari** [comment]
> after some thought it seems necessary. I was considering have a filter tab in which you could jump to a specific date from a calendar popup. The final design I'm envisioning will look something along the lines of:

We might also consider having separators in the scroll - e.g.

```
Jul 10 trip
Jul 10 trip
Jul 10 trip
---- July 11 ----
Jul 11 trip
Jul 11 trip
Jul 11 trip
```

**Comment by shankari** [comment]
```
console.log("Before reading from timeline, $scope.trip = "+$scope.trip);
console.log("using trip id "+stateParams.tripId+" to read the trip");
$scope.trip = ...
console.log("after reading from timeline, $scope.trip = "+$scope.trip);
```

**Comment by shankari** [comment]
Issue #1: why is `$scope.trip` set to undefined in your code? It is defined in the `multilabel-UI` which logs in the same way.
- check the matching between your controller name in the directive and the `.controller` definition. Because before this, your controller didn't actually do much so you may not have noticed if it wasn't linked up correctly.

`$scope.trip` should be defined in your controller since it is passed in

Next, look up what `$stateParams` does and how it is used.
HINT: as you can guess from the name, it has something to do with the parameters for a state.
Look at how we go to the detail screen. What function do we use, and how might it be related to `$stateParams`?

**Comment by shankari** [comment]
to debug, check how the multi-label-ui works
maybe try to pass in some other parameters and see if they work
basically figure out why `$scope.trip` doesn't get passed through

**Comment by tyleryandt18** [comment]
@shankari 
The investigation into why $scope.trip won't get passed through into the infinite_scroll_trip directive:

- Before reading from timeline, the trip is (well should be in this case) an object because it is passed into the directive. When the trip is a string, for example, 'Tyler rocks', that is what the trip will be before reading from the timeline.
- After reading from the timeline, at least in the case here that I sent you over teams, the getConfirmedTrip function returns undefined, and therefore sets the trip to undefined. The function has a conditional, where it either returns the trip from the timeline, or it returns undefined. The conditional statement is checking whether the trip with that given tripid is defined in the timeline.

And since our trip is mysteriously not being received, when we call that function, it sets the trip to be undefined because the trip object that is being passed in does not have a tripId associated with it. So undefined tripId in the function returns undefined.

**Comment by tyleryandt18** [comment]
Oh and when we are retrieving the trip, or attempting to, with the $stateParams tripId, there is nothing there. I'm not sure I understand why this would ever work, even if our trip was being passed in correctly. Is it because $stateParams has access to the attributes of the trip?

**Comment by shankari** [comment]
> Next, look up what $stateParams does and how it is used.
> HINT: as you can guess from the name, it has something to do with the parameters for a state.
> Look at how we go to the detail screen. What function do we use to go to the detail screen, and how might it be related to $stateParams?

**Comment by tyleryandt18** [comment]
While I'm doing that, could you touch briefly on if the code is using the detail.js file? I was looking into that file, and it goes about formatting its tripgj object ever so slightly, so I was going to see if pulling that code would work.

**Comment by shankari** [comment]
are you importing the module from detail.js? If not, we don't use anything from it.

**Comment by tyleryandt18** [comment]
I'm not importing the module, should I be? 

I was just noticing that there is a different method of retrieving the tripgj in that js file. And I was going to try out this code. But, I guess if we aren't using detail.js anymore then the code might be outdated.
<img width="707" alt="Screen Shot 2022-07-22 at 10 12 45 AM" src="https://user-images.githubusercontent.com/89606659/180480729-cb5ed074-2db1-4e97-a442-abb426da4b80.png">

This screenshot is from detail.js.

**Comment by shankari** [comment]
@tyleryandt18 you are going down the wrong path. Please answer my questions in 
https://github.com/e-mission/e-mission-docs/issues/740#issuecomment-1192713073
that we also discussed at our 1:1 yesterday

**Comment by tyleryandt18** [comment]
> Next, look up what $stateParams does and how it is used.
HINT: as you can guess from the name, it has something to do with the parameters for a state.
Look at how we go to the detail screen. What function do we use to go to the detail screen, and how might it be related to $stateParams?

$stateParams is a service that captures parameters and we can use it to display information according to the state. So when we go to a different state, the parameters would change. For example, the function that we use to go to the detail screen is a $state.go, and then it takes us to the different state and hence the different page. Within that function, we have a tripId: trip.id, so does that establish the tripId as a $stateParameter for that detail screen state?
@shankari

**Comment by shankari** [comment]
@tyleryandt18 yes. so what is the equivalent for the directive? What do we use the `tripId` for in the detail screen?

**Comment by tyleryandt18** [comment]
@shankari I'm not sure what you mean by the "equivalent for the directive". But no, we would not need to use $stateParams.tripId for that function. We could just do $scope.trip.id to go and retrieve the confirmed trip.
`$scope.trip = Timeline.getConfirmedTrip($stateParams.tripId);` -->
`$scope.trip = Timeline.getConfirmedTrip($scope.trip.id)`

**Comment by shankari** [comment]
Why do you need ` $scope.trip = Timeline.getConfirmedTrip($scope.trip.id)` when you already have `$scope.trip`?

**Comment by tyleryandt18** [comment]
> Why do you need $scope.trip = Timeline.getConfirmedTrip($scope.trip.id) when you already have $scope.trip?

Well now that is a great point. We wouldn't need it

**Comment by shankari** [comment]
Right. If you already have the trip, you just need to convert it to geojson to display it. Directly. Without lookup.
So you just need to figure out why the `$scope.trip` is `undefined` in the controller and you are done.

I don't know why `$scope.trip` is not defined in the controller. It seems like it should be given that it is a parameter to the directive. You need to poke at it with different options - you have already tried passing in a string - what if you pass in a different object.... The answer is bound to be something very stupid, probably a typo or an incorrect format issue.

**Comment by tyleryandt18** [comment]
@shankari
I just tried passing in a different object, and it was able to read it in fine. So it has specifically to do with the trip object, because when I set trip=a string, it was also able to read it in.

**Comment by shankari** [comment]
I know that `showDetail` worked for you earlier 
https://github.com/e-mission/e-mission-phone/pull/853/files#diff-338ab59c91e4724bfe144369b8f692359a4faa73acb869a0f6980ec6985c0cfdR32

if you try to `console.log` in there, is `$scope.trip` set?

**Comment by tyleryandt18** [comment]
It is only set after retrieving the trip with the tripid from $stateParams, but that makes sense
<img width="846" alt="Screen Shot 2022-07-22 at 11 54 30 AM" src="https://user-images.githubusercontent.com/89606659/180496636-d1a080e7-03b1-49a3-ae1a-d42e969e6039.png">
.

**Comment by shankari** [comment]
@tyleryandt18 we are not worried about the detail screen. we know it works. But when the user clicks on `showDetail`, it calls a function **in the directive**. Inside that function, is `$scope.trip` defined?

**Comment by tyleryandt18** [comment]
@shankari 
Right, sorry about that. So here now I added a console.log within the function, so when it's called it'll log. Logging the trip from there produces this result, so it is defined.
<img width="936" alt="Screen Shot 2022-07-22 at 12 01 23 PM" src="https://user-images.githubusercontent.com/89606659/180497646-6e2ac4ad-da78-44a0-bd00-1cac3d371b44.png">
`.showDetail()...{
console.log("Testing if showDetail has the trip defined ", $scope.trip)`

**Comment by tyleryandt18** [comment]
Also, to further prove that point, trip must be defined in order to see the map on the detail screen, which still works despite the map on the inf_scroll screen working. What I mean by this is that we pass the trip id into the $stateParams...

`tripId: trip.id` Within the showDetail function.
Therefore, it must be defined if goes to the detail screen and displays the correct state. This is so weird.

**Comment by shankari** [comment]
well, arguably, `tripId: trip.id` uses `trip` not `$scope.trip`.
But now I know what is going on and it is the async execution that I was talking about earlier.
Async operation is key to understanding javascript operation.

To understand what is going on, look at the ordering of the console logs.
- when is the controller created?
- where do the trips come from and when are they populated?
- when is console.log in `showDetails` invoked?

After you do all that, read up about `$apply` and `$watch`, which are techniques that angular uses to handle async execution.

**Comment by tyleryandt18** [comment]
>well, arguably, tripId: trip.id uses trip not $scope.trip.
Could you explain this just slightly? I'm not sure what the difference of this would be, because that showDetail function is a $scope function, is it not? So then to access the trip it would be the trip in the scope, and therefore the same as $scope.trip?

**Comment by shankari** [comment]
> because that showDetail function is a $scope function, is it not? So then to access the trip it would be the trip in the scope, and therefore the same as $scope.trip?

`showDetail` is a `$scope` function, but you also pass in `trip` as an argument to it. Inside the function, you currently use `trip`, which is the passed in argument.

To improve your directive implementation, you could remove that argument and change to `tripId: $scope.trip.id`

**Comment by tyleryandt18** [comment]
> When is the controller created?
So in the console logs, the infinite scroll controller is created before the trip item controller, the one for my directive. The Trip Item Ctrl is created fairly early on in the process.
<img width="350" alt="Screen Shot 2022-07-22 at 12 32 28 PM" src="https://user-images.githubusercontent.com/89606659/180502014-37aace21-b9d8-4868-b9ea-7eea84fd4422.png">




> Where do the trips come from and when are they populated?
Well the trips come from the server.
https://github.com/e-mission/e-mission-phone/blob/c0395cf3bf981e45257debad82c9c56e87dc4361/www/js/diary/infinite_scroll_list.js#L87
They are populated in the lines following that code, and therefore after the the controllers are called, correct? In the console, the trip item ctrl is called before the list of trips begins to be populated.

> When is console.log in showDetails() invoked?
The console.log in showDetails() is only invoked when I click on a button on the screen, so therefore after the controllers are called and the trip list is populated.

**Comment by shankari** [comment]
great! so what are some approaches for fixing it?
My high level thoughts are:
- `ng-repeat` vs `collection-repeat`
- `$watch`

You could also look into event broadcast/listen although IMHO that is not as attractive as the others

Please try out the first two options and list the pros and cons

**Comment by tyleryandt18** [comment]
> ng-repeat vs collection-repeat

I thought that we had to use collection-repeat in order for it to function on Android Devices?`     ` At least this is if you're talking about the pros and cons of switching it in the inf_scroll list html file.

**Comment by shankari** [comment]
look at the blame for that line - I wrote it in 2015.
So I would try and test it out on android and see if that still holds

**Comment by tyleryandt18** [comment]
@shankari I went ahead and switched the tripId: trip.id to tripId: $scope.trip, and the trip object is still defined when I log it to the console. Is there any significance to that?

**Comment by tyleryandt18** [comment]
> @shankari I went ahead and switched the tripId: trip.id to tripId: $scope.trip.id, and the trip object is still defined when I log it to the console. Is there any significance to that?

Does this mean, in theory, if I just created a $scope function that changed the trip object to a tripgj, that the trip object would be read in correctly? 

I'm going to get to the pros-cons list, I was just writing my write up so I'm just getting back on to this investigation.

**Comment by shankari** [comment]
> I went ahead and switched the tripId: trip.id to tripId: $scope.trip.id, and the trip object is still defined when I log it to the console. Is there any significance to that?
> Does this mean, in theory, if I just created a $scope function that changed the trip object to a tripgj, that the trip object would be read in correctly?

What do you think? I can tell you the answers, but the questions indicate that you haven't fully internalized what the various variables are and how they are set. Try it out, figure out the answers and then figure out the "why" behind the answers. You will understand better once you actually see it in action.

**Comment by tyleryandt18** [comment]
@shankari 
Pros and Cons List
ng-repeat vs collections repeat
- First off, I am not too sure how to run an android emulator. I've only been working on Xcode and the apple version. I probably could just go read the phone ReadME tho huh. Anyways, I think I would also have to install android studio through IT, so the ng-repeat test on Android might have to wait until Monday. But, I will try it out and make sure to update you on this thread.
- Wow ok I find this really interesting. The ng-repeat calls the Trip Item Controller after populating the list of trips. In the console, the order of logs goes Infinite Scroll Diary Ctrl called, list populating, THEN trip item ctrl called. As opposed to the collection-repeat has the trip item ctrl called before populating the list.
- YOOOOOOOOOOOO IT WORKED!!!!!!! wow that made me happy. I just uncommented the part that actually creates the map after using ng-repeat, and displayed the maps correctly. 
- Downside tho, and I feel as if this would be an easy fix, within the collection repeat, there is the html tag for the start and end time buttons, so when we switch that to ng-repeat, there is only the start time tag for the very first trip, and the end time tag for the very last trip. We could easily just change up the directive html template to include these time tags there, and then we would be set.

**Comment by shankari** [comment]
why don't you go ahead with fixing the easy fix for the downside then, and make the changes to have the directive look like the preferred mockup at the same time? Add screenshots once you are done.

**Comment by shankari** [comment]
@tyleryandt18 has now finished the UI directive changes and will add them to the PR pretty soon along with TODOs.

In terms of reading from the server, we currently read data in 

```
Timeline.readAllConfirmedTrips(currEnd, ONE_WEEK).then((ctList) => {
```

and we append it to the current list 

```
        $scope.data.allTrips = ctList.concat($scope.data.allTrips);
```

So the way that we load the data is:
- load first week (list has 10 elements)
- load week before (list has 20 elements)
- load week before that (list has 30 elements)

....

**Comment by shankari** [comment]
- when the user selects a date from a year ago, we just pass in that date to `readAllConfirmedTrips`

 Issue:
- what if they select a date that they don't have a trip for?
- make sure that if they select a date that they have already selected, we don't read it twice
- don't allow them to go to a date in the future
- what they go to a year ago and then try to scroll down? Right now we start at the end, and we append to the current list, so there is always data on the phone when we scroll down, until we get to the end. But if we jump to a year ago, then the data from the time between a year ago and now will not be retrieved to the phone. So we won't have it when the user scrolls down.

**Comment by tyleryandt18** [comment]
> * make sure that if they select a date that they have already selected, we don't read it twice
> * what they go to a year ago and then try to scroll down? Right now we start at the end, and we append to the current list, so there is always data on the phone when we scroll down, until we get to the end. But if we jump to a year ago, then the data from the time between a year ago and now will not be retrieved to the phone. So we won't have it when the user scrolls down.

@shankari I feel as if these two issues/solutions are conflicting. When a user jumps to a date from a year ago, are we allowing them to scroll up and down? I don't believe so, so therefore when they select a date, it just displays that date's trips, correct? So then why are we not allowing them to click on a date that they have already selected? if they use the date selector to go from a year ago, then go to a date like yesterday, why could they not use the date selector to go back to that date in the past?

**Comment by tyleryandt18** [comment]
@shankari Furthermore, I don't know if I will be able to complete the date selector functionality. I really don't understand it, and I feel as if there might be a solution to the UI even better than the date selector that Sebastian could implement. If you don't believe that the trip draft function is working, then maybe I should focus on making sure that works along with the trip percentages, because I am also stumped on that implementation, but am confident I can work it out.

**Comment by shankari** [comment]
> When a user jumps to a date from a year ago, are we allowing them to scroll up and down? I don't believe so, so therefore when they select a date, it just displays that date's trips, correct?

That is not what I assumed. So you are saying that if the user selects a date, they only get to see data for that date? But otherwise, they can scroll up and down?

That does not seem like a great user experience. If we allow people to "infinite scroll", they will assume that they can scroll all the time. I recall you saying that you want to remove the buttons near the calendar date selector because people will just scroll. Without scrolling, we run into all the original issues about knowing which dates actually have data and so on.

**Comment by shankari** [comment]
Closing this given that we have merged @tyleryandt18's changes and now have a label screen that is fairly close to the diary and opening https://github.com/e-mission/e-mission-docs/issues/902 for the pending changes.
